File: src/main/java/com/company/apimgmt/service/SystemService.java
package com.company.apimgmt.service;

import com.company.apimgmt.dto.CreateSystemRequest;
import com.company.apimgmt.dto.SystemDTO;
import com.company.apimgmt.dto.UpdateSystemRequest;
import com.company.apimgmt.entity.SystemEntity;
import com.company.apimgmt.exception.DuplicateResourceException;
import com.company.apimgmt.exception.ResourceNotFoundException;
import com.company.apimgmt.exception.ValidationException;
import com.company.apimgmt.repository.SystemRepository;
import io.smallrye.mutiny.Uni;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@ApplicationScoped
public class SystemService {

    @Inject
    SystemRepository systemRepository;

    public Uni<SystemDTO> createSystem(CreateSystemRequest request) {
        // Validate system name uniqueness
        return systemRepository.findByName(request.name)
                .onItem().transformToUni(existing -> {
                    if (existing != null) {
                        return Uni.createFrom().failure(
                                new DuplicateResourceException("System", "name", request.name));
                    }

                    // Create new system entity
                    SystemEntity entity = new SystemEntity();
                    entity.id = UUID.randomUUID();
                    entity.name = request.name;
                    entity.description = request.description;
                    entity.createdAt = LocalDateTime.now();
                    entity.updatedAt = LocalDateTime.now();

                    return systemRepository.create(entity)
                            .onItem().transform(this::toDTO);
                });
    }

    public Uni<SystemDTO> getSystemById(UUID id) {
        return systemRepository.findById(id)
                .onItem().transform(entity -> {
                    if (entity == null) {
                        throw new ResourceNotFoundException("System", id);
                    }
                    return toDTO(entity);
                });
    }

    public Uni<List<SystemDTO>> getAllSystems() {
        return systemRepository.findAll()
                .onItem().transform(entities -> entities.stream()
                        .map(this::toDTO)
                        .collect(Collectors.toList()));
    }

    public Uni<SystemDTO> updateSystem(UUID id, UpdateSystemRequest request) {
        return systemRepository.findById(id)
                .onItem().transformToUni(entity -> {
                    if (entity == null) {
                        return Uni.createFrom().failure(
                                new ResourceNotFoundException("System", id));
                    }

                    // Check name uniqueness if name is being updated
                    if (request.name != null && !request.name.equals(entity.name)) {
                        return systemRepository.findByName(request.name)
                                .onItem().transformToUni(existing -> {
                                    if (existing != null) {
                                        return Uni.createFrom().failure(
                                                new DuplicateResourceException("System", "name", request.name));
                                    }
                                    return updateSystemEntity(entity, request);
                                });
                    } else {
                        return updateSystemEntity(entity, request);
                    }
                });
    }

    private Uni<SystemDTO> updateSystemEntity(SystemEntity entity, UpdateSystemRequest request) {
        if (request.name != null) {
            entity.name = request.name;
        }
        if (request.description != null) {
            entity.description = request.description;
        }
        entity.updatedAt = LocalDateTime.now();

        return systemRepository.update(entity)
                .onItem().transform(this::toDTO);
    }

    public Uni<Void> deleteSystem(UUID id) {
        return systemRepository.findById(id)
                .onItem().transformToUni(entity -> {
                    if (entity == null) {
                        return Uni.createFrom().failure(
                                new ResourceNotFoundException("System", id));
                    }

                    // Check for cascade constraints - ensure no APIs reference this system
                    return systemRepository.countApisBySystemId(id)
                            .onItem().transformToUni(count -> {
                                if (count > 0) {
                                    return Uni.createFrom().failure(
                                            new ValidationException(
                                                    String.format(
                                                            "Cannot delete system. %d API(s) are still associated with this system",
                                                            count)));
                                }

                                return systemRepository.delete(id)
                                        .onItem().transform(deleted -> null);
                            });
                });
    }

    private SystemDTO toDTO(SystemEntity entity) {
        return new SystemDTO(
                entity.id,
                entity.name,
                entity.description,
                entity.createdAt,
                entity.updatedAt);
    }
}

File: apimgmt/services/system_service.py
from typing import List, Optional
import uuid
from datetime import datetime

from apimgmt.repositories.system_repository import SystemRepository
from apimgmt.services.audit_service import AuditLogService
from apimgmt.models.system import System
from apimgmt.schemas.system import CreateSystemRequest, UpdateSystemRequest, SystemDTO
from apimgmt.exceptions import ResourceNotFoundException, DuplicateResourceException, ValidationException
from apimgmt.utils.audit_decorator import audit_log
from apimgmt.models.audit import OperationType, ResourceType


class SystemService:
    def __init__(self, system_repository: SystemRepository, audit_service: Optional[AuditLogService] = None):
        self.system_repository = system_repository
        self.audit_service = audit_service
        # 打印调试信息
        import logging
        logging.info(f"SystemService initialized with audit_service: {audit_service is not None}")
    
    @audit_log(
        operation_type=OperationType.CREATE,
        resource_type=ResourceType.SYSTEM
    )
    def create_system(self, request: CreateSystemRequest) -> SystemDTO:
        """创建系统"""
        # 检查系统名称是否已存在
        existing_system = self.system_repository.find_by_name(request.name)
        if existing_system:
            raise DuplicateResourceException("System", "name", request.name)
        
        # 检查系统编号是否已存在
        existing_system_by_code = self.system_repository.find_by_system_code(request.system_code)
        if existing_system_by_code:
            raise DuplicateResourceException("System", "system_code", request.system_code)
        
        # 创建新系统
        system = System(
            id=str(uuid.uuid4()),
            name=request.name,
            system_code=request.system_code,
            description=request.description,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        created_system = self.system_repository.create(system)
        
        return SystemDTO.model_validate(created_system)
    
    def get_system_by_id(self, system_id: str) -> SystemDTO:
        """根据ID获取系统"""
        system = self.system_repository.find_by_id(system_id)
        if not system:
            raise ResourceNotFoundException("System", system_id)
        return SystemDTO.model_validate(system)
    
    def get_all_systems(self) -> List[SystemDTO]:
        """获取所有系统"""
        systems = self.system_repository.find_all()
        return [SystemDTO.model_validate(system) for system in systems]
    
    @audit_log(
        operation_type=OperationType.UPDATE,
        resource_type=ResourceType.SYSTEM,
        resource_id_param="system_id"
    )
    def update_system(self, system_id: str, request: UpdateSystemRequest) -> SystemDTO:
        """更新系统"""
        # 查找系统
        system = self.system_repository.find_by_id(system_id)
        if not system:
            raise ResourceNotFoundException("System", system_id)
        
        # 检查名称唯一性（如果更新名称）
        if request.name and request.name != system.name:
            existing_system = self.system_repository.find_by_name(request.name)
            if existing_system:
                raise DuplicateResourceException("System", "name", request.name)
        
        # 更新系统
        if request.name:
            system.name = request.name
        if request.description is not None:
            system.description = request.description
        system.updated_at = datetime.utcnow()
        
        updated_system = self.system_repository.update(system)
        
        return SystemDTO.model_validate(updated_system)
    
    @audit_log(
        operation_type=OperationType.DELETE,
        resource_type=ResourceType.SYSTEM,
        resource_id_param="system_id"
    )
    def delete_system(self, system_id: str) -> None:
        """删除系统"""
        # 查找系统
        system = self.system_repository.find_by_id(system_id)
        if not system:
            raise ResourceNotFoundException("System", system_id)
        
        # 检查是否有关联的API
        api_count = self.system_repository.count_apis_by_system_id(system_id)
        if api_count > 0:
            raise ValidationException(
                f"Cannot delete system. {api_count} API(s) are still associated with this system"
            )
        
        # 删除系统
        self.system_repository.delete(system_id)
